File: AXAPI/HaoConnect/php/HaoConnect/HaoResult.php
<?php
/**
 * user接口返回的数据转换成Result实例，用于支持点语法输出数据以及各种筛选
 * @package Model
 * @author axing
 * @since 0.1
 * @version 0.1
 */
class HaoResult {

	/* 特殊变量，没有set和get方法，不推荐外部使用。 */
    public $errorCode;
    public $errorStr;
    public $extraInfo;
    public $results;
    public $modelType;
    public $searchIndexString;

    public $pathCache;

    // ======================== functions ========================

    /** 将数组初始化成对象 */
    public static function instanceModel($results,$errorCode,$errorStr,$extraInfo)
    {

        $modelType = 'HaoResult';

        if (is_array($results))
        {
            if (isset($results['modelType']))
            {
                $modelType = $results['modelType'];
            }
        }

	    $modelName = $modelType . ( $modelType != 'HaoResult' ? 'Result' : '' );
	    if (class_exists($modelName))
	    {
	        $object = new $modelName();
	    }
	    else
	    {
	    	$object = new HaoResult();
	    }

        $object ->errorCode   = $errorCode;
        $object ->errorStr    = $errorStr;
        $object ->extraInfo   = $extraInfo;
        $object ->results     = $results;
        $object ->modelType   = $modelType;

        $object ->pathCache   = array();

        return $object;
    }


    /** 根据路径取数据，默认是results中取，也可以指定extraInfo>路径下取数据。 */
    public function find($path)
    {
        $path = trim($path);

        if (isset($this->pathCache[$path]))
        {
            return $this->pathCache[$path];
        }

        if ( strpos($path,'results>') !== 0 && strpos($path,'extraInfo>') !== 0 )
        {
            $path = 'results>'.$path;
        }

        $paths = explode('>',$path);


        $changeValue = null;

        foreach ($paths as $index=>$keyItem) {
            if ($index==0)
            {
                if ($keyItem=='extraInfo')
                {
                    $changeValue = $this->extraInfo;
                }
                else
                {
                    $changeValue = $this->results;
                }
            }
            else if ($keyItem!='')
            {
                if (isset($changeValue[$keyItem]))
                {
                    $changeValue = $changeValue[$keyItem];
                    continue;
                }
                $changeValue = null;
                break;
            }
        }

        $value = $this->value($changeValue);
        $this->pathCache[$path] = $value;
        return $value;
    }


    /** 传入值如果是model，则以当前Result为框架构建新Result，否则直接返回。 */
    public function value($value)
    {
        if (is_array($value))
        {
            if (isset($value['modelType']))
            {
                return static::instanceModel($value, $this->errorCode, $this->errorStr, $this->extraInfo, $this->resultCount);
            }
            else
            {
                $result = array();
                foreach ($value as $key => $tmpValue)
                {
                    $result[$key] = $this->value($tmpValue);
                }
                $value = $result;
            }
        }
        return $value;
    }

    /**
     * 根据path取值，如果不是数组，就转成数组
     * @param  string $path
     * @return array
     */
    public function findAsList($path)
    {
        $value = $this->find($path);

        if (!is_array($value) || (array_keys($value) !== array_keys(array_keys($value))) )
        {
            $value = array($value);
        }

        return $value;
    }

    /**
     * 根据path取值，如果不是字符串，就转成字符串
     * @param  string $path
     * @return string
     */
    public function findAsString($path)
    {
        $value = $this->find($path);

        if (!is_string($value))
        {
            $value = json_encode($value,JSON_UNESCAPED_UNICODE);
        }

        return $value;
    }


    /**
     * 根据path取值，如果不是数字，就转成数字
     * @param  string $path
     * @return int
     */
    public function findAsInt($path)
    {
        $value = $this->find($path);

        if (!is_int($value))
        {
            $value = intval(json_encode($value,JSON_UNESCAPED_UNICODE));
        }

        return $value;
    }

    /**
     * 根据path取值，如果不是HaoResult类型，就转成HaoResult类型
     * @param  string $path
     * @return HaoResult
     */
    public function findAsResult($path)
    {
        $value = $this->find($path);

        if (!is_subclass_of($value,'HaoResult'))
        {
            $value = HaoResult::instanceModel($value, $this->errorCode, $this->errorStr, $this->extraInfo, $this->resultCount);
        }

        return $value;
    }


    /** 在结果中进行搜索，返回结果是数组（至少也是空数组） */
    public function search($path)
    {
        if ($this->searchIndexString == null)
        {
            $resultsIndex            = HaoUtility::getKeyIndexArray( ['results'=>$this->results] );
            $extraInfoIndex          = HaoUtility::getKeyIndexArray( ['extraInfo'=>$this->extraInfo] );
            $searchIndex             = array_merge( $resultsIndex , $extraInfoIndex );
            $this->searchIndexString = implode("\n",$searchIndex);
        }

        $path = trim($path);

        if ( strpos($path,'results>') !== 0 && strpos($path,'extraInfo>') !== 0 )
        {
        	$path = 'results>'.$path;
        }

        $result = array();
        $searchFoundCount = preg_match_all('/(^|\s)('.$path.')\s+/',$this->searchIndexString,$matches);
        if (is_array($matches) && count($matches)>=2)
        {
            foreach ($matches[2] as $pathMatched) {
                $result[] = $this->find($pathMatched);
            }
        }

        return $result;

    }

    /** 判断当前实例是否目标model */
    public function isModelType($modelType)
    {
        return strtolower($modelType) == strtolower($this->modelType);
    }

    /**
     * 判断是否等于目标ErroCode
     * @param  array  $errorCode  目标errorCode
     * @return boolean            是否一致
     */
    public function isErrorCode($errorCode)
    {
        return $this->errorCode === $errorCode;
    }

    /**
     * 判断是否正确获得结果
     * @return boolean            是否正确获得
     */
    public function isResultsOK()
    {
        return $this->isErrorCode(0) ;
    }

    /** 返回字典类型数据（重新包装成字典） */
    public function properties()
    {
        return array(
                        'errorCode'  => $this->errorCode,
                        'errorStr'   => $this->errorStr,
                        'extraInfo'  => $this->extraInfo,
                        'results'    => $this->results
                    );
    }

}

File: AXAPI/HaoConnect/java/HaoConnect/HaoResult.java
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import java.lang.reflect.Array;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Created by wangtao on 15/12/21.
 */
public class HaoResult {

    public int errorCode = 0;
    public String errorStr;
    public JsonElement extraInfo;
    public JsonElement results;
    public String modelType;
    public String searchIndexString;

    public Map<String, Object> pathCache;

    public static Object instanceModel(JsonElement results, int errorCode, String errorStr, JsonElement extraInfo) {

        String modelType = "HaoResult";

        if (results != null && results instanceof JsonObject) {
            if (((JsonObject) results).get("modelType") != null) {
                modelType = ((JsonObject) results).get("modelType").getAsString();
            }
        }

        String resultName = HaoResult.class.getPackage().getName() + ".HaoResult";

        if (!modelType.equals("HaoResult")) {
            resultName = modelType + "Result";
            resultName = HaoResult.class.getPackage().getName() + ".results." + resultName;
        }

        Class c;
        HaoResult objc = null;
        try {
            c = Class.forName(resultName);
            Constructor constructor = c.getDeclaredConstructor();
            constructor.setAccessible(true);

            objc = (HaoResult) constructor.newInstance();
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
            objc = new HaoResult();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        } catch (InstantiationException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        }

        objc.errorCode = errorCode;
        objc.errorStr = errorStr;
        objc.extraInfo = extraInfo;
        objc.results = results;
        objc.modelType = modelType;

        objc.pathCache = new HashMap<>();

        return objc;
    }

    public Object find(String path) {

        path = path.trim();

        if (pathCache.containsKey(path)) {
            return pathCache.get(path);
        } else {
            if (!path.startsWith("results>") && !path.startsWith("extraInfo>")) {
                path = "results>" + path;
            }

            String[] paths = path.split(">");

            JsonElement changeValue = null;

            for (int i = 0; i < paths.length; i++) {
                if (i == 0) {
                    if (paths[0].equals("extraInfo")) {
                        changeValue = this.extraInfo;
                    } else {
                        changeValue = this.results;
                    }
                } else {
                    String keyItem = paths[i];
                    if (keyItem.length() > 0) {
                        if (keyItem.matches("\\d+")) {
                            if (changeValue instanceof JsonArray) {
                                if (Integer.parseInt(keyItem) <= ((JsonArray) changeValue).size()) {
                                    changeValue = ((JsonArray) changeValue).getAsJsonArray().get(Integer.parseInt(keyItem));
                                    continue;
                                }
                            }
                        } else {
                            if (changeValue instanceof JsonObject) {
                                changeValue = ((JsonObject) changeValue).get(keyItem);
                                continue;
                            }
                        }
                        changeValue = null;
                        break;
                    }
                }
            }

            Object value = value(changeValue);
            pathCache.put(path, value);
            return value;
        }

    }

    public Object value(JsonElement value) {
        if (value instanceof JsonObject) {
            if (((JsonObject) value).get("modelType") != null) {
                return HaoResult.instanceModel(value, this.errorCode, this.errorStr, this.extraInfo);
            }
        } else if (value instanceof JsonArray) {
            ArrayList<Object> array = new ArrayList<>();
            for (int i = 0; i < ((JsonArray) value).size(); i++) {
                JsonElement objc = (JsonElement) ((JsonArray) value).get(i);
                array.add(value(objc));
            }
            return array;
        }
        return value;
    }

    /**
     * @param path
     * @return
     */
    public ArrayList<Object> findAsList(String path) {
        Object objc = find(path);
        if (!(objc instanceof JsonArray || objc instanceof List || objc instanceof Array)) {
            ArrayList<Object> list = new ArrayList<>();
            list.add(objc);
            return list;
        } else {
            return (ArrayList<Object>) objc;
        }
    }

    public String findAsString(String path) {
        try {
            return find(path).toString();
        } catch (Exception e) {
        }
        return "";
    }

    public int findAsInt(String path) {
        try {
            return Integer.parseInt(find(path).toString());
        } catch (Exception e) {

        }
        return -1;
    }

    public Object findAsResult(String path) {
        Object objc = find(path);
        if (!(objc instanceof HaoResult)) {
            if (objc instanceof JsonElement) {
                return instanceModel((JsonElement) objc, this.errorCode, this.errorStr, this.extraInfo);
            }
        }
        return objc;
    }

    public ArrayList<Object> search(String path) {
        if (this.searchIndexString == null) {
            JsonObject resultObjc = new JsonObject();
            resultObjc.add("results", this.results);
            ArrayList<String> resultsIndex = HaoUtility.getKeyIndexArray(resultObjc);

            JsonObject extraInfoObjc = new JsonObject();
            extraInfoObjc.add("extraInfo", this.extraInfo);
            ArrayList<String> extraIndex = HaoUtility.getKeyIndexArray(extraInfoObjc);
            resultsIndex.addAll(extraIndex);

            ArrayList<String> searchIndex = resultsIndex;

            StringBuffer tempIndex = new StringBuffer();
            for (int i = 0; i < searchIndex.size(); i++) {
                tempIndex.append(searchIndex.get(i) + "\n");
            }
            HaoUtility.print(tempIndex.toString());
            this.searchIndexString = tempIndex.toString();
        }

        path = path.trim();

        if (!path.startsWith("results>") && !path.startsWith("extraInfo>")) {
            path = "results>" + path;
        }

        ArrayList<Object> result = new ArrayList<>();

        String regEx = "(^|\\s)(" + path + ")\\s+";
        Pattern pat = Pattern.compile(regEx);
        Matcher mat = pat.matcher(this.searchIndexString);
        while (mat.find()) {
            result.add(find(mat.group()));
        }

        return result;
    }

    public boolean isModelType(String modelType) {
        return this.modelType.equalsIgnoreCase(modelType);
    }

    public boolean isErrorCode(int errorCode) {
        return this.errorCode == errorCode;
    }

    public boolean isResultsOK() {
        return isErrorCode(0);
    }
}
